src/main/java/com/google/sps/servlets/TripServlet.java

```diff
@@ -18,16 +33,155 @@
+        return findPlaceResult.candidates[0].placeId;
+      }
+      
+      // No candidate is given, so return null.
+      return null;
+    } catch(ApiException | InterruptedException e) {
+      throw new IOException(e);
     }
   }
 
+  /**
+   * Get the PlaceDetails object from the place ID.
+   */
+  public PlaceDetails getPlaceDetailsFromPlaceId(GeoApiContext context, String placeId)
+    throws IOException {
+
+    PlaceDetailsRequest placeDetailsRequest = PlacesApi.placeDetails(context, 
+      placeId);
+    try {
+      return placeDetailsRequest.await();
+    } catch(ApiException | InterruptedException e) {
+      throw new IOException(e);
+    }
+  }
+
+  /**
+   * Populate the destinationName and photoSrc fields using the Google Maps API.
+   */
+  public void populateDestinationAndPhoto(GeoApiContext context, String tripDestination)
+    throws IOException {
+
+    // Get place ID from search of trip destination. Get photo and destination 
+    // if not null; otherwise, use a placeholder photo and destination.
+    String destinationPlaceId = getPlaceIdFromTextSearch(context, this.tripDestination);
+    if (destinationPlaceId == null) {
+      this.destinationName = tripDestination;
+      this.photoSrc = "../images/placeholder_image.png";
+    } else {
+      PlaceDetails placeDetailsResult = getPlaceDetailsFromPlaceId(context, destinationPlaceId);
+
+      // Get the name of the location from the place details result.
+      this.destinationName = placeDetailsResult.name;
+
+      // Get a photo of the location from the place details result.
+      if (placeDetailsResult.photos == null) {
+        this.photoSrc = "../images/placeholder_image.png";
+      } else {
+        Photo photoObject = placeDetailsResult.photos[0];
+        this.photoSrc = getUrlFromPhotoReference(400, photoObject.photoReference);
+      }
+    }
+  }
+
+  /**
+   * Store the Trip Entity in datastore with the User Entity as an ancestor.
+   * Return the Trip Entity object.
+   */
+  public Entity storeTripEntity(HttpServletResponse response, String tripName, 
+    String destinationName, String tripDayOfTravel, String photoSrc) throws IOException {
+    // Get User Entity. If user not logged in, redirect to homepage.
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    Entity userEntity = AuthServlet.getCurrentUserEntity();
+    if (userEntity == null) {
+      response.sendRedirect("/");
+      return null;
+    }
+
+    // Put Trip Entity into datastore.
+    Entity tripEntity = Trip.buildEntity(tripName, destinationName, photoSrc,
+      tripDayOfTravel, tripDayOfTravel, userEntity.getKey());
+    datastore.put(tripEntity);
+    return tripEntity;
+  }
+
+  /**
+   * Get a URL to show the photo from the photoreference.
+   * See https://developers.google.com/places/web-service/photos#place_photo_requests
+   * for more info.
+   */
+  public String getUrlFromPhotoReference(int maxWidth, String photoReference) {
```

Comment:
Could you add a description of the parameters in the comment? I'm not sure what these params store

Reply:
Done!